### ISSUE TITLE
Python does not evaluate type annotations on assignments in function bodies

### ISSUE THREAD

**Issue by finswimmer** [snok/flake8-type-checking#186]
Hey,

`ruff` update its `UP037` rule (See: https://github.com/astral-sh/ruff/pull/11414), which means that quotes around type hints in a function body are not necessary (I couldn't find a source for this, but it seems to work).

With this change this code:

```py
def foo() -> None:
    bar: "Bar"
```

becomes

```py
def foo() -> None:
    bar: Bar
```

and `flake8-type-checking` starts to complain "TC200 Annotation 'Bar' needs to be made into a string literal" of course.

Any chance to fix this?

Thanks a lot. :pray: 

fin swimmer

**Comment by sondrelg** [comment]
Hmm yes at first glance this seems safe - and would definitely be an improvement. I have some other commitments I need to get to before I'll have time to implement this, but think I would be happy to accept a PR in the meantime 👍

**Comment by Daverball** [comment]
We technically don't know for sure that it's safe, since it depends on when the function is first called, so we shouldn't emit an error in either case (quoted or unquoted). So we would need some special casing that only applies if we find the symbol in an outer scope, instead of our own.

**Comment by sondrelg** [comment]
Can you think of an example where it would it not be safe? Running this seems to work well:

```python
def foo():
    x: pandas.DataFrame

foo()
```

**Comment by Daverball** [comment]
Well the import can happen inside an `if TYPE_CHECKING:` block, then it is unsafe no matter what, but it can also happen below the call-site of `foo`:

```python
def foo():
    x: pandas.DataFrame

foo()

import pandas
```

**Comment by Daverball** [comment]
Alright, I think I misunderstood the problem. I didn't even know Python already always defers annotations inside function bodies. So we should treat anything within a function scope as if we had a `from __future__ import annotations` import, i.e. any wrapped annotation should be an error and any unwrapped annotation shouldn't be one.

**Comment by Daverball** [comment]
On second thought we should probably just special case all `AnnAssign` as long as we're inside a function scope, since we don't even need to check whether the symbol exists or not. We can keep a separate list of wrapped annotations that should always be unwrapped.

**Comment by finswimmer** [comment]
That was fast. :rocket: Thanks a lot :partying_face:
